source/common/http/filter_chain_helper.cc

```diff
@@ -0,0 +1,153 @@
+#include "source/common/http/filter_chain_helper.h"
+
+#include <memory>
+#include <string>
+
+#include "envoy/registry/registry.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/common/fmt.h"
+#include "source/common/config/utility.h"
+#include "source/common/http/utility.h"
+#include "source/common/protobuf/utility.h"
+#include "source/extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+namespace Http {
+
+class MissingConfigFilter : public Http::PassThroughDecoderFilter {
+public:
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override {
+    decoder_callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::NoFilterConfigFound);
+    decoder_callbacks_->sendLocalReply(Http::Code::InternalServerError, EMPTY_STRING, nullptr,
+                                       absl::nullopt, EMPTY_STRING);
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+};
+
+static Http::FilterFactoryCb MissingConfigFilterFactory =
+    [](Http::FilterChainFactoryCallbacks& cb) {
+      cb.addStreamDecoderFilter(std::make_shared<MissingConfigFilter>());
+    };
+
+void FilterChainHelper::createFilterChainForFactories(
+    Http::FilterChainManager& manager,
+    const FilterChainHelper::FilterFactoriesList& filter_factories) {
+  bool added_missing_config_filter = false;
+  for (const auto& filter_config_provider : filter_factories) {
+    auto config = filter_config_provider->config();
+    if (config.has_value()) {
+      Filter::NamedHttpFilterFactoryCb& factory_cb = config.value().get();
+      manager.applyFilterFactoryCb({filter_config_provider->name(), factory_cb.name},
+                                   factory_cb.factory_cb);
+      continue;
+    }
+
+    // If a filter config is missing after warming, inject a local reply with status 500.
+    if (!added_missing_config_filter) {
+      ENVOY_LOG(trace, "Missing filter config for a provider {}", filter_config_provider->name());
+      manager.applyFilterFactoryCb({}, MissingConfigFilterFactory);
+      added_missing_config_filter = true;
+    } else {
+      ENVOY_LOG(trace, "Provider {} missing a filter config", filter_config_provider->name());
+    }
+  }
+}
+
+void FilterChainHelper::processFilters(const FiltersList& filters, const std::string& prefix,
+                                       const std::string& filter_chain_type,
+                                       FilterFactoriesList& filter_factories) {
+  DependencyManager dependency_manager;
+  for (int32_t i = 0; i < filters.size(); i++) {
```

Comment:
I could be wrong, but it looks like Protobuf::RepeatedPtrField::size() returns int, not int32_t. Should we use int to avoid truncation? (Not that we're likely to have more than 2^32 fields)